### DISCUSSION TITLE
Avoiding the cost of "fixed" when a struct is known to be situated in unmanaged memory.

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#8011] [Language Ideas]
Methods in structs that need to get their own address (`&this`) currently can only do so within a `fixed` statement. This _could_ be avoided if the `struct` were somehow annotated in such a way that it could never be situated in the managed heap.

Perhaps `unmanaged` or `fixed` or something, could be added as an option when declaring a `struct`. Any attempt to declare a field of this type inside a `class` (or nested inside other structs inside a class) would fail. Because the `struct` can be guaranteed to never be situated in the managed heap, it will be safe to allow methods and properties in these structs to avoid the overhead of using `fixed` when getting their own address.

Thoughts?

**Comment by HaloFour** [comment]
Isn't that already `ref struct`?

**Reply by The-Futurist** [reply]
No, that isn't enough. we get a compile error in a ref struct if we try to get our own address. Also I don't think you can use a ref struct as a type argument.

**Reply by HaloFour** [reply]
I meant that we don't need new syntax to indicate a struct that cannot be on the heap.  That's exactly what `ref struct` means.  I would agree that it doesn't mean that the compiler already supports what you're suggesting, but I would think that such behavior could be built on top of `ref struct` vs. a new concept that means the same thing.

> Also I don't think you can use a ref struct as a type argument.

Yet.  Any other mechanism for expressing that the struct can't be on the heap would have the exact same restriction, you'd need that anti-constraint to convey that the generic method is safe and never attempts to box the struct.

**Reply by The-Futurist** [reply]
Yes, I see, so perhaps then a first step would be to eliminate the need for using `fixed` to get a pointer, if the type being pointed to is a `ref struct` that's a good start. Then perhaps `ref` as a new constraint would have some merits...

**Reply by CyrusNajmabadi** [reply]
While the restriction could be lifted, I don't really see much motivation.  Writing a 'fixed' statement is trivial, so it's not a substantial burden to keep the status quo here.

**Reply by The-Futurist** [reply]
But isn't there a performance cost, needless too in this scenario?

**Reply by CyrusNajmabadi** [reply]
I don't see why there would be. Ref structs shouldn't move. What performance issue are you observing? If there is one, the runtime would be the right place to investigate and do something, not the language.

**Reply by The-Futurist** [reply]
No data, I just assumed that there is a cost and of one needs to do this a lot that cost would accrue. That aside, this surely is a language issue because it _insists_ on the developer doing something that is pointless and unnecessary - or so it seems to me.

**Reply by CyrusNajmabadi** [reply]
Absent data, I don't see any need to change things here. It would be high cost for marginal (or zero) gain.

> or so it seems to me.

Language changes are enormously costly. We don't do them unless we think there is substantial gain to the ecosystem as a whole. Having to write this construct is on the order of a paper cut. Investments in something like that doesn't really make sense given all the other areas we can focus on. 

If you have an actual perf issue here, I would bring it up with the runtime first so we can see what they think. If this cannot be solved there and there is substantive need, we can look at the lang or compiler side. The latter also seems fine to me. The lang could require this, but the compiler could no-op it

**Comment by The-Futurist** [comment]
Also, unsafe code cannot appear inside iterators, if this is related to the possibility of the GC messing up the iterators state, then could this too be changed so that if the unsafe code is only accessing a `ref struct` it's safe to allow iterators?

Eric Lipper explains that (apart from it being in bad taste) it is because of the GC and the need to perhaps pin something for the duration of the enumeration, so again if the unsafe code is only accessing a `ref struct` this could be allowed - in principle. 

Imagine a scenario where we need to walk raw memory, perhaps an unmanaged linked list or something, this is a use case.

**Reply by huoyaoyuan** [reply]
It's almost the opposite. Don't think async methods to be methods, they are state machines.

Direct fixed pointer is additional task for GC.  Indirect pins can already be achived via GCHandle. Ref struct with state machine is fundamentally incompatible.

**Reply by The-Futurist** [reply]
@huoyaoyuan 

See:

https://github.com/dotnet/roslyn/issues/2143
https://github.com/dotnet/csharplang/discussions/510

**Comment by The-Futurist** [comment]
If we have a `fixed` field array and I simply want to enumerate it, expose it through some public property, I cannot do so.

**Reply by CyrusNajmabadi** [reply]
What stops you in this case?  Can you link to the code you're using to try this to do this?
